noria-server/mir/src/optimize.rs

```diff
@@ -1,19 +1,217 @@
     //remove_extraneous_projections(&mut q);
-    q
+    find_and_merge_filter_aggregates(&mut q)
 }
 
 pub fn optimize_post_reuse(_q: &mut MirQuery) {
     // find_and_merge_filter_chains(q);
 }
 
+fn find_and_merge_filter_aggregates(q: &mut MirQuery) -> Vec<MirNodeRef> {
+
+    // 1. depth first search to find all the nodes, so we can process them later
+
+    let mut node_stack = Vec::new();
+    node_stack.extend(q.roots.iter().cloned());
+
+    let mut visited_nodes = HashMap::new();
+    let mut found_nodes = Vec::new();
+
+    while !node_stack.is_empty() {
+        let n = node_stack.pop().unwrap();
+        let node_name = n.borrow().versioned_name();
+
+        if visited_nodes.contains_key(&node_name) {
+            continue;
+        }
+
+        for child in n.borrow().children.iter() {
+            node_stack.push(child.clone());
+        }
+
+        visited_nodes.insert(node_name, true);
+        found_nodes.push(n);
+    }
+
+    // 2. iterate over the nodes to find candidates, i.e.
+    // nodes that are a filter followed by an aggregate with no other children/parents.
+    // TODO (jamb) Initially we also considered nodes that are an aggregate followed by a filter,
+    // but this case is rarer and may have the problem that it is filtering on the result
+    // of the aggregate, in which case they cannot be merged.
+
+    let mut candidate_nodes = Vec::new();
+
+    for n in found_nodes {
+        // if we've already set this to false, we shouldn't look at it more;
+        // e.g. it might be merging with its parent node.
+        let node_name = n.borrow().versioned_name();
+        if !visited_nodes.get(&node_name).unwrap() {
+            continue;
+        }
+        // now scan for candidacy
+        let mut candidate = false;
+        match n.borrow().inner {
+            MirNodeType::Filter { .. }  => {
+                // if there's exactly one child and it's an aggregation and it has exactly one parent,
+                // then this is a candidate
+                if n.borrow().children.len() == 1 {
+                    let temp = n.borrow();
+                    let child = temp.children.first().unwrap().borrow();
+                    match child.inner {
+                        MirNodeType::Aggregation { .. } => {
+                            if child.ancestors.len() == 1 {
+                                candidate = true;
+                            }
+                        },
+                        _ => {},
+                    };
+                }
+            },
+            MirNodeType::Aggregation { ref on, .. }  => {
+                // if there's exactly one child and it's an aggregation and it has exactly one parent,
+                // then this is a candidate
+                if n.borrow().children.len() == 1 {
+                    let temp = n.borrow();
+                    let child = temp.children.first().unwrap().borrow();
+                    match child.inner {
```

Comment:
This can be `if let MirNodeType::Filter { ref conditions } = child.inner` to save a level of indentation